### PR TITLE
Hide over-exposed public APIs part 2

### PR DESCRIPTION
**Pull request by minwoox**
- (Breaking)
  - Move `common.util.NonNullByDefault` to `common.annotation.NonNullByDefault`
  - Move `common.util.UnstableApi` to `common.annotation.UnstableApi`
  - `KeyedCircuitBreakerMapping` and `KeyedCircuitBreakerMapping.KeySelector` are now package-private.
    - Use static methods in `CircuitBreakerMapping`.
- (Deprecated)
  - `ClientFactoryOptions.of()` is deprecated. Use `ClientFactoryOptions.empty()`.
  - `MetricCollectingCircuitBreakerListener` is deprecated.
     Use `CircuitBreakerListener.metricCollecting()`.
  - `RateLimitingThrottlingStrategy` is deprecated. Use `ThrottlingStrategy#rateLimiting()`.
    
- Miscellaneous
  - Add rnn and final where applicable
  - Update Javadoc

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/MetricCollectingCircuitBreakerListener.java [hunk 1]
```diff
@@ -57,22 +57,28 @@
  */
 public final class MetricCollectingCircuitBreakerListener implements CircuitBreakerListener {
```
[the remark this answers is not in the document]
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/circuitbreaker/MetricCollectingCircuitBreakerListener.java:61]: Fixed. Thanks. 😉

File: core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerMapping.java [hunk 2]
```diff
@@ -71,7 +61,7 @@ static CircuitBreakerMapping perHost(Function<String, CircuitBreaker> factory) {
      *                creates a new {@link CircuitBreaker}
      */
     static CircuitBreakerMapping perHostAndMethod(Function<String, CircuitBreaker> factory) {
-        return new KeyedCircuitBreakerMapping<>(KeySelectorUtil.hostAndMethodSelector, factory);
+        return new KeyedCircuitBreakerMapping<>(KeySelector.HOST_AND_METHOD, factory);
     }
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerMapping.java]: How about making it accept a `BiFunction`?

**Comment by trustin** [comment]
Oh, and the `(Bi)Function`s could produce `? extends CircuitBreaker` instead of `CircuitBreaker`.

**Comment by trustin** [comment]
Could you fix the compilation error so I can merge? 😉

**Comment by minwoox** [comment]
Oops fixed. 😉

**Comment by trustin** [comment]
One step closer to 1.0 :tada:

**Pull request merged**